gtsam/hybrid/tests/testIncrementalHybrid.cpp

```diff
@@ -45,296 +44,334 @@ using symbol_shorthand::X;
+  GaussianHybridFactorGraph graph1;
+
+  // Add the 3 DC factors, x1-x2, x2-x3, x3-x4
+  for (size_t i = 0; i < 3; i++) {
+    graph1.push_back(switching.linearizedFactorGraph.dcGraph().at(i));
+  }
+
+  // Add the Gaussian factors, 1 prior on X(1), 4 measurements
+  for (size_t i = 0; i <= 4; i++) {
+    graph1.push_back(switching.linearizedFactorGraph.gaussianGraph().at(i));
+  }
+
+  // Create ordering.
+  Ordering ordering;
+  for (size_t j = 1; j <= 4; j++) {
+    ordering += X(j);
+  }
+
+  // Now we calculate the actual factors using full elimination
+  HybridBayesNet::shared_ptr unprunedHybridBayesNet;
+  GaussianHybridFactorGraph::shared_ptr unprunedRemainingGraph;
+  std::tie(unprunedHybridBayesNet, unprunedRemainingGraph) =
+      switching.linearizedFactorGraph.eliminatePartialSequential(ordering);
+
+  size_t maxComponents = 5;
+  incrementalHybrid.update(graph1, ordering, maxComponents);
+
+  /*
+  unpruned factor is:
+      Choice(m3)
+      0 Choice(m2)
+      0 0 Choice(m1)
+      0 0 0 Leaf 0.2248 -
+      0 0 1 Leaf 0.3715 -
+      0 1 Choice(m1)
+      0 1 0 Leaf 0.3742 *
+      0 1 1 Leaf 0.6125 *
+      1 Choice(m2)
+      1 0 Choice(m1)
+      1 0 0 Leaf 0.3706 -
+      1 0 1 Leaf 0.6124 *
+      1 1 Choice(m1)
+      1 1 0 Leaf 0.611 *
+      1 1 1 Leaf    1 *
+
+  pruned factors is:
+      Choice(m3)
+      0 Choice(m2)
+      0 0 Leaf    0
+      0 1 Choice(m1)
+      0 1 0 Leaf 0.3742
+      0 1 1 Leaf 0.6125
+      1 Choice(m2)
+      1 0 Choice(m1)
+      1 0 0 Leaf    0
+      1 0 1 Leaf 0.6124
+      1 1 Choice(m1)
+      1 1 0 Leaf 0.611
+      1 1 1 Leaf    1
+   */
+
+  // Test if the remaining factor graph has one factor and if it is the expected
+  // discrete factor.
+  auto remainingFactorGraph = incrementalHybrid.remainingFactorGraph();
+  EXPECT_LONGS_EQUAL(1, remainingFactorGraph.size());
+
+  auto discreteFactor_m1 = *dynamic_pointer_cast<DecisionTreeFactor>(
+      incrementalHybrid.remainingDiscreteGraph().at(0));
+  EXPECT(discreteFactor_m1.keys() == KeyVector({M(3), M(2), M(1)}));
+
+  // Get the number of elements which are equal to 0.
+  auto count = [](const double &value, int count) {
+    return value > 0 ? count + 1 : count;
+  };
+  EXPECT_LONGS_EQUAL(5, discreteFactor_m1.fold(count, 0));
+
+  /* Expected hybrid Bayes net
+   * factor 0:  [x1 | x2 m1 ], 2 components
+   * factor 1:  [x2 | x3 m2 m1 ], 4 components
+   * factor 2:  [x3 | x4 m3 m2 m1 ], 8 components
+   * factor 3:  [x4 | m3 m2 m1 ], 8 components
+   */
+  auto hybridBayesNet = incrementalHybrid.hybridBayesNet();
+
+  EXPECT_LONGS_EQUAL(4, hybridBayesNet.size());
+  EXPECT_LONGS_EQUAL(2, hybridBayesNet.atGaussian(0)->nrComponents());
+  EXPECT_LONGS_EQUAL(4, hybridBayesNet.atGaussian(1)->nrComponents());
+  EXPECT_LONGS_EQUAL(5, hybridBayesNet.atGaussian(2)->nrComponents());
+  EXPECT_LONGS_EQUAL(5, hybridBayesNet.atGaussian(3)->nrComponents());
+
+  auto &lastDensity = *(hybridBayesNet.atGaussian(3));
```

Comment:
why de-reference?

Reply:
So we can call `lastDensity(assignment)` instead of `lastDensity->operator()(assignment)`.